rust-mozjs/examples/wasm.rs

```diff
@@ -0,0 +1,177 @@
+//! (environment within one global object) before you can execute code.
+
+use std::ptr;
+use std::ptr::null_mut;
+
+#[macro_use]
+extern crate mozjs;
+use mozjs::jsapi::*;
+use mozjs::jsval::ObjectValue;
+use mozjs::jsval::UndefinedValue;
+use mozjs::rust::jsapi_wrapped::{Construct1, JS_GetProperty, JS_SetProperty};
+use mozjs::rust::SIMPLE_GLOBAL_CLASS;
+use mozjs::rust::{JSEngine, RealmOptions, Runtime};
+
+/// hi.wat:
+/// ```
+/// (module
+///  (import "env" "bar" (func $bar (param i32) (result i32)))
+///  (func (export "foo") (result i32)
+///    i32.const 42
+///    call $bar
+///  ))
+///```
+const HI_WASM: [libc::c_uchar; 56] = [
+    0x00, 0x61, 0x73, 0x6d, 0x01, 0x00, 0x00, 0x00, 0x01, 0x0a, 0x02, 0x60, 0x01, 0x7f, 0x01, 0x7f,
+    0x60, 0x00, 0x01, 0x7f, 0x02, 0x0b, 0x01, 0x03, 0x65, 0x6e, 0x76, 0x03, 0x62, 0x61, 0x72, 0x00,
+    0x00, 0x03, 0x02, 0x01, 0x01, 0x07, 0x07, 0x01, 0x03, 0x66, 0x6f, 0x6f, 0x00, 0x01, 0x0a, 0x08,
+    0x01, 0x06, 0x00, 0x41, 0x2a, 0x10, 0x00, 0x0b,
+];
+
+unsafe extern "C" fn bar(_cx: *mut JSContext, argc: u32, vp: *mut Value) -> bool {
+    let args = CallArgs::from_vp(vp, argc);
+    args.rval().set(args.get(0).get());
+    true
+}
+
+fn run(rt: Runtime) {
+    let options = RealmOptions::default();
+    rooted!(in(rt.cx()) let global = unsafe {
+        JS_NewGlobalObject(rt.cx(), &SIMPLE_GLOBAL_CLASS, ptr::null_mut(),
+                           OnNewGlobalHookOption::FireOnNewGlobalHook,
+                           &*options)
+    });
+    let _ac = JSAutoRealm::new(rt.cx(), global.get());
+
+    // Get WebAssembly.Module and WebAssembly.Instance constructors.
+    rooted!(in(rt.cx()) let mut wasm = UndefinedValue());
+    rooted!(in(rt.cx()) let mut wasm_module = UndefinedValue());
+    rooted!(in(rt.cx()) let mut wasm_instance = UndefinedValue());
+
+    unsafe {
+        assert!(JS_GetProperty(
+            rt.cx(),
+            global.handle(),
+            b"WebAssembly\0".as_ptr() as *const libc::c_char,
+            &mut wasm.handle_mut()
+        ));
+        rooted!(in(rt.cx()) let mut wasm_obj = wasm.to_object());
+        assert!(JS_GetProperty(
+            rt.cx(),
+            wasm_obj.handle(),
+            b"Module\0".as_ptr() as *const libc::c_char,
+            &mut wasm_module.handle_mut()
+        ));
+        assert!(JS_GetProperty(
+            rt.cx(),
+            wasm_obj.handle(),
+            b"Instance\0".as_ptr() as *const libc::c_char,
+            &mut wasm_instance.handle_mut()
+        ));
+
+        // Construct Wasm module from bytes.
+        rooted!(in(rt.cx()) let mut module = null_mut::<JSObject>());
+        {
+            let array_buffer = JS::NewArrayBufferWithUserOwnedContents(
+                rt.cx(),
+                HI_WASM.len(),
+                HI_WASM.as_ptr() as _,
+            );
+            assert!(!array_buffer.is_null());
+
+            let args = HandleValueArray::from_rooted_slice(&[ObjectValue(array_buffer)]);
```

Comment:
This slice is not actually rooted. I'll try to provide useful suggestions later when I'm not on my phone.

Reply:
Would something like this be ok:
```rust
rooted!(in(rt.cx()) let val = ObjectValue(array_buffer));
let args = HandleValueArray::from_rooted_slice(&[val.get()]);
```

Reply:
Hmm, almost, but not quite. There's an awkward but correct way of doing that:
```rust
rooted!(in(rt.cx()) let val = ObjectValue(array_buffer));
let args = HandleValueArray { length_: 1, elements_: &*val };
```
In Servo we have the RootedVec type which could probably be used to build a cleaner binding here, but that's not part of this crate.